src/silo/storage/column/insertion_index.cpp

```diff
@@ -0,0 +1,235 @@
+   // The candidate insertions are selected based on the 3-mers within the search pattern.
+   // If an insertion id is in all candidate sets, then we have a viable candidate which
+   // might match the search pattern. If this does not hold, there exists a continuous
+   // 3-mer which is in the search pattern but not in the candidate insertion. Therefore, the
+   // regex will never match and we can ignore this sequence.
+
+   using it = insertion_ids_t::const_iterator;
+   std::vector<std::pair<it, it>> min_heap;
+   for (const auto& three_mer : search_three_mers) {
+      const auto& candidate_insertions =
+         three_mer_index.at(three_mer[0]).at(three_mer[1]).at(three_mer[2]);
+      if (candidate_insertions.empty()) {
+         continue;
+      }
+      min_heap.emplace_back(candidate_insertions.cbegin(), candidate_insertions.cend());
+   }
+
+   if (min_heap.size() < search_three_mers.size()) {
+      return std::make_unique<sequence_ids_t>();
+   }
+
+   const auto cmp = [](const std::pair<it, it>& lhs_it_pair, const std::pair<it, it>& rhs_it_pair) {
+      return *lhs_it_pair.first > *rhs_it_pair.first;
+   };
+   std::make_heap(min_heap.begin(), min_heap.end(), cmp);
+
+   auto result = std::make_unique<sequence_ids_t>();
+
+   size_t count = 0;
+   uint32_t current_insertion_id = *min_heap.front().first;
+   while (!min_heap.empty()) {
+      const auto next_insertion_id = *min_heap.front().first;
+
+      std::pop_heap(min_heap.begin(), min_heap.end(), cmp);
+      ++min_heap.back().first;
+
+      if (min_heap.back().first == min_heap.back().second) {
+         min_heap.pop_back();
+      } else {
+         std::push_heap(min_heap.begin(), min_heap.end(), cmp);
+      }
+
+      if (next_insertion_id != current_insertion_id) {
+         if (count == search_three_mers.size()) {
+            const auto& insertion = insertions[current_insertion_id];
+            if (std::regex_search(insertion.value, search_pattern)) {
+               *result |= insertion.sequence_ids;
+            }
+         }
+         count = 1;
+         current_insertion_id = next_insertion_id;
+      } else {
+         ++count;
+      }
+   }
+
+   if (count == search_three_mers.size()) {
+      const auto& insertion = insertions[current_insertion_id];
+      if (std::regex_search(insertion.value, search_pattern)) {
+         *result |= insertion.sequence_ids;
+      }
+   }
+
+   return result;
+}
+
+void InsertionIndex::addLazily(const std::string& insertions_string, uint32_t sequence_id) {
+   if (insertions_string.empty()) {
+      return;
+   }
+   for (auto& position_and_insertion : splitBy(insertions_string, DELIMITER_INSERTIONS)) {
+      auto [position, insertion] = parseInsertion(position_and_insertion);
+
+      auto it1 =
+         collected_insertions.emplace(position, std::unordered_map<std::string, sequence_ids_t>{});
+      auto it2 = it1.first->second.emplace(insertion, sequence_ids_t{});
```

Comment:
Please improve the naming. Maybe something along the lines of:
```suggestion
      auto insertions_at_position =
         collected_insertions.emplace(position, std::unordered_map<std::string, sequence_ids_t>{}).first;
      auto sequence_ids_at_position = insertions_at_position->second.emplace(insertion, sequence_ids_t{});
```

Reply:
Adapted it